### PR TITLE
feat: eslint rules

### PR DESCRIPTION
**Pull request by rebelchris**
Introducing our very own custom color check for ESLint 🎉
The idea is that this ESLint plugin will check for the tailwind config and go through all valid bg and text color attributes.

This will be V1 and has no particular test yet.
It can also be abstracted way more.

DD-277 #done

### CODE REVIEW

File: packages/eslint-config/index.js [hunk 1]
```diff
@@ -1,42 +1,43 @@
 module.exports = {
-  parser: '@typescript-eslint/parser',
+  parser: "@typescript-eslint/parser",
```
Comment (idoshamun) [inline packages/eslint-config/index.js:2]: Oh noes! Please use single quote

File: packages/eslint-config/package.json [hunk 2]
```diff
@@ -3,6 +3,7 @@
   "version": "0.0.0",
   "main": "index.js",
   "devDependencies": {
+    "eslint-plugin-daily": "*",
```
Comment (idoshamun) [inline packages/eslint-config/package.json:6]: Why is this different than this `@dailydotdev/eslint-plugin-daily-dev-eslint-rules`?

File: packages/eslint-rules/README.md [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# `eslint-rules`
```
Comment (idoshamun) [inline packages/eslint-rules/README.md:1]: Not needed at this point unless you want to document something

File: packages/eslint-rules/lib/index.js [hunk 4]
```diff
@@ -0,0 +1,3 @@
+module.exports.rules = {
+  "no-custom-color": require("./rules/no-custom-color"),
```
Comment (idoshamun) [inline packages/eslint-rules/lib/index.js:2]: Single quote please

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 5]
```diff
@@ -0,0 +1,174 @@
+const tailwindConfig = require("../../../shared/tailwind.config");
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:1]: Single quote everywhere please 😉

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+  meta: {
+    type: "problem",
+    docs: {
+      description: "Incorrect command class name",
+      suggestion: false,
+    },
+    fixable: "code",
+  },
+  create: (context) => {
+    const targetClassNames = ["bg-", "text-"];
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:22]: What about the border and shadow props?

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+    const findTargetClassNames = (classNames) => {
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:23]: I think we can put it outside the `create` function for better readability

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+      }
+
+      return classNames.filter((className) => {
+        return targetClassNames.some((matchClassName) =>
+          className.startsWith(matchClassName)
+        );
+      });
+    };
+
+    const mapValidClassNames = (colors, prefix = "") => {
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:47]: I think we can put it outside the `create` function for better readability

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+        if (typeof colors[k] === "object" && colors[k] !== null) {
+          const newPrefix = prefix ? `${prefix}-${k}` : k;
+          r = [
+            ...new Set(r.concat(r, mapValidClassNames(colors[k], newPrefix))),
+          ];
+        } else {
+          r.push(`${prefix && `${prefix}-`}${k}`);
+        }
+        return r;
+      }, []);
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:58]: This feels quite hard to read for me. I don't feel like I understand 100% this code

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    const mappedValidClassNames = [
+      ...alwaysValidClasses,
+      ...mapValidClassNames(theme.colors).flatMap((color) => {
+        return targetClassNames.flatMap((prefix) => {
+          return `${prefix}${color}`;
+        });
+      }),
+    ];
+
+    const isClassAttribute = (node) => {
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:71]: I think we can put it outside the `create` function for better readability

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+      return node.name && /^class(Name)?$/.test(node.name.name);
+    };
+
+    const isLiteralAttributeValue = (node) => {
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:75]: I think we can put it outside the `create` function for better readability

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+          node,
+          message: `${className} is not a theme color`,
+          data: {
+            classname: className,
+          },
+        });
+      });
+    };
+
+    const isValidJSXAttribute = (node) => {
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:111]: I think we can put it outside the `create` function for better readability

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+        return false;
+      }
+      if (!isLiteralAttributeValue(node)) {
+        // No support for dynamic or conditional classnames
+        return false;
+      }
+      return true;
+    };
+
+    const getClassNameValuesFromNode = (node) => {
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:123]: I think we can put it outside the `create` function for better readability

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+      if (node.value?.type === "JSXExpressionContainer") {
+        return node.value.expression.value;
+      }
+
+      return node.value?.value;
+    };
+
+    const parseNodeRecursive = (node, arg, callback) => {
```
Comment (idoshamun) [inline packages/eslint-rules/lib/rules/no-custom-color.js:131]: I think we can put it outside the `create` function for better readability

File: packages/eslint-rules/package.json [hunk 15]
```diff
@@ -0,0 +1,26 @@
+{
+  "name": "eslint-plugin-daily",
```
Comment (idoshamun) [inline packages/eslint-rules/package.json:2]: Do we miss the scope here? `@dailydotdev`

**Review by idoshamun — CHANGES_REQUESTED** [review]
This looks like quite a work here! You learned super fast how to create eslint plugins 🎉

**Review by sshanzel — CHANGES_REQUESTED** [review]
You've got great work going on here! 🚀 

Just had a few very minor non-blocker comments.

File: packages/eslint-rules/lib/rules/no-custom-color.js (hunk 6 above)
Comment (rebelchris) [reply packages/eslint-rules/lib/rules/no-custom-color.js:22]: We can add those, but I think the default tailwind plugin actually already checks for those.

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+    docs: {
+      description: "Incorrect command class name",
+      suggestion: false,
+    },
+    fixable: "code",
+  },
+  create: (context) => {
+    const targetClassNames = ["bg-", "text-"];
+    const findTargetClassNames = (classNames) => {
+      if (classNames === undefined) {
```
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/eslint-rules/lib/rules/no-custom-color.js:24]: Makes total sense!

File: packages/eslint-rules/lib/rules/no-custom-color.js [hunk 17]
```diff
@@ -0,0 +1,174 @@
+const tailwindConfig = require("../../../shared/tailwind.config");
+const alwaysValidClasses = [
+  "text-base",
+  "text-center",
+  "text-2xl",
+  "text-left",
+  "text-lg",
+  "text-xl",
```
[the remark this answers is not in the document]
Comment (rebelchris) [reply packages/eslint-rules/lib/rules/no-custom-color.js:8]: They are used quite widely for icons ;)

[the remark this answers is not in the document]
Comment (sshanzel) [reply packages/eslint-rules/lib/rules/no-custom-color.js]: @idoshamun should we update those implementations to follow our guidelines?

**Comment by rebelchris** [comment]
@idoshamun / @sshanzel thanks guys!
I've refactored the readability.

The main function is still quite big, so would love to hear if it makes more sense with the new naming.

**Review by idoshamun — APPROVED** [review]
Looks great!

**Review by sshanzel — APPROVED** [review]

**Pull request merged**